capreolus/trainer/__init__.py

```diff
@@ -292,3 +297,240 @@ def predict(self, reranker, pred_data, pred_fn):
+            "query": tf.train.Feature(float_list=tf.train.FloatList(value=query)),
+            "query_idf": tf.train.Feature(float_list=tf.train.FloatList(value=query_idf)),
+            "posdoc_id": tf.train.Feature(bytes_list=tf.train.BytesList(value=[posdoc_id.encode('utf-8')])),
+            "posdoc": tf.train.Feature(float_list=tf.train.FloatList(value=posdoc)),
+        }
+
+        if negdoc_id:
+            feature["negdoc_id"] = tf.train.Feature(bytes_list=tf.train.BytesList(value=[negdoc_id.encode('utf-8')])),
+            feature["negdoc"] = tf.train.Feature(float_list=tf.train.FloatList(value=negdoc))
+
+        return feature
+
+    def write_tf_record_to_file(self, dir_name, tf_features):
+        filename = dir_name / "{}.tfrecord".format(str(uuid.uuid4()))
+        examples = [tf.train.Example(features=tf.train.Features(feature=feature)) for feature in tf_features]
+
+        def generator():
+            for example in examples:
+                yield example.SerializeToString()
+
+        dataset = tf.data.Dataset.from_generator(generator, output_types=tf.string)
+        if not os.path.isdir(dir_name):
+            os.makedirs(dir_name, exist_ok=True)
+
+        writer = tf.data.experimental.TFRecordWriter(str(filename))
+        writer.write(dataset)
+
+        return str(filename)
+
+    def convert_to_tf_dev_record(self, dataset):
+        dir_name = self.get_cache_path() / dataset.get_hash()
+
+        tf_features = [
+            self.create_tf_feature(sample["qid"], sample["query"], sample["query_idf"], sample["posdocid"], sample["posdoc"], None, None)
+            for sample in dataset
+        ]
+
+        return [self.write_tf_record_to_file(dir_name, tf_features)]
+
+    def convert_to_tf_train_record(self, dataset):
+        """
+        Tensorflow works better if the input data is fed in as tfrecords
+        Takes in a pytorch IterableDataset, iterates through it, and creates a tf record from it
+        The randomization/shuffling e.t.c is handled by IterableDataset. See sampler/__init__.py
+        """
+        tf_record_filenames = []
+        tf_features = []
+        dir_name = self.get_cache_path() / dataset.get_hash()
+
+        # There are 'n' iterations
+        # Each iterations has a size 'itersize' number of batches in it
+        for niter in tqdm(range(0, self.cfg["niters"]), desc="Converting data to tf records"):
+            for sample_idx, data in enumerate(dataset):
+                # TODO: Split into multiple files. This will destroy the RAM
+                tf_features.append(
+                    self.create_tf_feature(
+                        data["qid"], data["query"], data["query_idf"], data["posdocid"], data["posdoc"],
+                        data["negdocid"], data["negdoc"]
+                    )
+                )
+                if sample_idx + 1 >= self.cfg["itersize"] * self.cfg["batch"]:
```

Comment:
I probable need a better strategy here. What I do now:

1. Iterate through the dataset to get `niters * itersize * batch` number of samples
2. Save the ^ above samples to a tf record
3. When I make use of the tf record, I can do a simple iteration (i.e `for x in tf_record`) without bothering about itersize or niters since the tf records already contains the exact number of samples I need.
4. This might pose problems for caching. Eg: I create tf records and cache it for niters=10, itersize=32, batch=16. Next time I train the tf records will be read from the cache and if I set my batch size as 32 this time, it would mean that I would not have enough training samples in my tf record (assuming I'd still want niters=10 and itersize=32).

Reply:
Two possible solutions:

1. We could make the cache key a function of the sampling parameters and the total number of samples `total_samples = niters * itersize * batch`. Then when looking for a cached tfrecord, we load any that has <= the number of samples we need.

2. We could write the unique samples out once and then use [tf.data.Dataset.shuffle](https://www.tensorflow.org/api_docs/python/tf/data/Dataset#shuffle) to sample from them. This is what Crystina's TFv1 implementation does. If we go for this option, we need to be careful to make sure `shuffle`'s `buffer_size` is large enough that we see new data every iteration. When writing the unique samples, we still need to be careful to iterate over them in the right order (e.g., round robin over queries).

The second option is more straightforward but reduces our control over sampling. I think this is okay though.

Reply:
Hmmmm so for option 2, you are suggesting that we think in terms of epochs right? An epoch in this case would be `num_queries * num_posdocs_in_qrels * num_negdocs_in_qrels`. So all we have to do is save one epoch and then keep shuffling it every time we load it? This sounds like the better approach